### ISSUE TITLE
Adding a footer to all pages

### ISSUE THREAD

**Issue by chathurag** [Cacti/cacti#1203]
Is there any way to add a footer to all the pages of cacti including login page?

**Comment by netniV** [comment]
Currently, this could be done using a plugin (if one existed) as the bottom_footer() function does raise an api call for two functions:

```
api_plugin_hook_function('bottom_footer', OPER_MODE_NATIVE);
api_plugin_hook('page_bottom');
```
The first can be used to determine if a footer should be displayed, the second adds footer details.  If you want to avoid using the plugin methods, a mod could be made to check for the existence of a custom footer and include it, though how the display of that looks may vary depending on which theme is used by the system.

Also note, that with a customer mod, any plugin installed may not obey the cacti display format so may never call bottom_footer()

**Comment by cigamit** [comment]
Can you provide me an example of what you are looking for?

**Comment by chathurag** [comment]
Need to add company logo and some text for all the pages.

**Comment by cigamit** [comment]
You should be able to hack that into a theme if you like.  Check the paper-plane theme.  He has added a scroll area at the bottom of each page.  You should be able to extend that to your own theme(s).

If not, the plugin, maybe the old loginmod plugin on the doco site might be the way to go as previously indicated if you need something quick.  You will have to update to version 1.0 rules though.

**Comment by chathurag** [comment]
In simply I just need to add Company Logo & IPv6 Enabled Logo with hyperlink to the footer of each page on classisc theme... Is it possible?

**Comment by cigamit** [comment]
Best done with a simple plugin to do this if you need it any time soon.  Could be done using javascript as was pointed out earlier.  Also, as netniV pointed out, you could do it using that bottom footer hook.  There are several plugin examples you can use.

**Comment by chathurag** [comment]
Could you please guide me to do this?

**Comment by xmacan** [comment]
I will do it. I prefer to do it directly in Cacti (settings->Visual) and not with a plugin. Why another plugin for this small thing? There will be checkbox for enable/disable footer and textarea with few allowed html tags.Is it ok?

**Comment by cigamit** [comment]
The Plugin concept is perfect for small little changes that not many people will use, which I believe this falls under that category.  You would still see it in Settings > Visual, the plugin would just put the setting there.

**Comment by xmacan** [comment]
You are right, I will prepare small plugin. I need few days.

**Comment by xmacan** [comment]
I have prepared first version of Plugin footer - https://github.com/xmacan/plugin_footer 

I found one problem -  on graph page (tree mode) I have footer more times.

![screenshot_2018-07-10 tree mode -](https://user-images.githubusercontent.com/26485719/42534375-dec19e66-848c-11e8-9771-9e76b5346107.png)

**Comment by netniV** [comment]
I'll check out the plugin and then try it out.  I'll let you know what I find.

**Comment by netniV** [comment]
How did you get the footer display?  When I attempt this under 1.2, I am seeing that the footer div is off the page.

**Comment by xmacan** [comment]
I have 1.1.38. I'm using page_bottom hook. It is concept only, tested only with classic theme.
I don't understand "footer div is off the page". Please show screenshot.

If footer isn't displayed, please try console -> settings -> visual, check "Display footer" and save.

**Comment by netniV** [comment]
The problem comes with the styling of the "non" classic themes.  These seem to set heights etc. on the divs I think so force the div you insert to be outside of the window.  I am not sure how to force it to push the other div up (reduce its height) using just CSS without some javascript at the moment.

**Comment by netniV** [comment]
If you log the number of times you are called, is it more than once?

**Comment by netniV** [comment]
The reason I ask is that I suspect that this is the issue of an ajax call vs a normal page load.  In a normal page load, then the header/footer would be wanted.  In an ajax call, no headers or footers are normally requested.

I think in 1.2 we go a stage further and actually only extract the `#cacticontent` area but I'd have to double check to be sure. So that would be self-retify the issue.  But you could also check for the "header" option and suppress the footer if "false"

**Comment by xmacan** [comment]
On graph page tree mode is called twice
2018/07/11 10:39:07 - footer Footer plugin displayed 0.34169200 1531298347
2018/07/11 10:39:09 - footer Footer plugin displayed 0.47734100 1531298349

**Comment by netniV** [comment]
Can you try adding the stack trace to that?  You should be able to use the following:

```php
clean_up_lines(var_export(debug_backtrace(), true))
```

Although, clean_up_lines may be a 1.2 function, I can't quite remember.  If it is, just steal the function from the source and rename it slightly into your plugin for testing purposes.

**Comment by netniV** [comment]
The other thing I am wondering is if this is due to flow maybe going as follows:

1. Load page
2. Apply filters
3. Load ajax call

So, I would again suggest checking for the header value and suppressing if its false.

**Comment by xmacan** [comment]
with var_export I found:
'file' => '/usr/local/share/cacti/graph_view.php', 'line' => 193, 'function' => 'bottom_footer' (tree)
'file' => '/usr/local/share/cacti/graph_view.php', 'line' => 328, 'function' => 'bottom_footer' (tree content)

I guess that second call is unnecessary.

Condition if (!isset_request_var('header') || get_nfilter_request_var('header') == 'false') didn't help me. It is fulfilled in both cases.

**Comment by netniV** [comment]
Check for action=tree_content maybe then?

**Comment by xmacan** [comment]
Yes, it works. But I'm trying to say that the second call in graph_view.php (line 328) is unnecessary. Or isn't?

**Comment by netniV** [comment]
Probably.

**Comment by xmacan** [comment]
I found problem (or maybe it is my mistake). In my plugin (https://github.com/xmacan/plugin_footer) I would like add file upload button for upload company logo for footer. But form in setting.php hasn't multipart/form-data. I tried hardcode it, but still I cannot upload anything:
 PHP Notice:  Undefined index: plugin_footer_logo in /usr/local/share/cacti/plugins/footer/setup.php
And I don't know why. Is it possible work with $_FILES variables?
templates_import.php works with $_FILES. Where is problem, please?

**Comment by netniV** [comment]
Check thaf yoy have weire access ro your tmo folder as the websire user. That stops it working sometimes. If you are still having problems i can test it out for you.

**Comment by netniV** [comment]
Corrected the above message that my phone managed to garble really well!

**Comment by cigamit** [comment]
Not to jump in late to this thing, but you could always do this in almost 100% jQuery with a callback to gather the message from the database.  There is a div already designed for a bottom footer, you just have to make it visible as @browniebraun has done with a few of his themes, and insert the text into that section, even scroll it if that makes sense.

**Comment by xmacan** [comment]
Now I released first version of plugin. It is very small thing and now supports only classic themes. I will continue with themes support, logo upload, ....

 https://github.com/xmacan/plugin_footer

**Comment by cigamit** [comment]
Congrats!!!
